randovania/generator/elevator_distributor.py

```diff
@@ -164,85 +160,46 @@ def create_elevator_database(region_list: RegionList,
-            if elevators.mode == TeleporterShuffleMode.ECHOES_SHUFFLED:
-                connections = elevator_echoes_shuffled(game, rng)
-
-            elif elevators.mode in {TeleporterShuffleMode.TWO_WAY_RANDOMIZED, TeleporterShuffleMode.TWO_WAY_UNCHECKED}:
-                connections = two_way_elevator_connections(
-                    rng=rng,
-                    elevator_database=elevator_db,
-                    between_areas=elevators.mode == TeleporterShuffleMode.TWO_WAY_RANDOMIZED
-                )
-            else:
-                connections = one_way_elevator_connections(
-                    rng=rng,
-                    elevator_database=elevator_db,
-                    target_locations=elevators.valid_targets,
-                    replacement=elevators.mode != TeleporterShuffleMode.ONE_WAY_ELEVATOR,
-                )
-
-            elevator_connection.update(connections)
-
-        for teleporter, destination in elevators.static_teleporters.items():
-            elevator_connection[teleporter] = destination
-
-        assignment = [
-            (region_list.typed_node_by_identifier(identifier, DockNode), region_list.node_by_identifier(target))
-            for identifier, target in elevator_connection.items()
-        ]
-
-        return assignment
-
-def elevator_echoes_shuffled(game_description: GameDescription, rng: Random) -> ElevatorConnection:
-        from randovania.games.prime2.generator.base_patches_factory import WORLDS
-        worlds = list(WORLDS)
-        rng.shuffle(worlds)
-
-        result = {}
-
-        def area_to_node(identifier: AreaIdentifier):
-            area = game_description.region_list.area_by_area_location(identifier)
-            for node in area.actual_nodes:
-                if node.valid_starting_location:
-                    return node.identifier
-            raise KeyError(f"{identifier} has no valid starting location")
-
-        def link_to(source: AreaIdentifier, target: AreaIdentifier):
-            result[area_to_node(source)] = area_to_node(target)
-            result[area_to_node(target)] = area_to_node(source)
-
-        def tg_link_to(source: str, target: AreaIdentifier):
-            link_to(AreaIdentifier("Temple Grounds", source), target)
+    region_list = game.region_list
+    elevator_connection: ElevatorConnection = {}
+
+    if not teleporters.is_vanilla:
+        if rng is None:
+            raise MissingRng("Elevator")
+
+        elevator_dock_types = game.dock_weakness_database.all_teleporter_dock_types
+        elevator_db = create_elevator_database(region_list, teleporters.editable_teleporters,
+                                                                    elevator_dock_types)
+        if teleporters.mode == TeleporterShuffleMode.ECHOES_SHUFFLED:
+            from randovania.games.prime2.generator.elevator_distributor import elevator_echoes_shuffled
+            connections = elevator_echoes_shuffled(game, rng)
```

Comment:
hm... I don't like this wild echoes-specific thing here.
Maybe just error completely if this happens, and Echoes' base patch factory does the check and redirects instead.

Reply:
Is there an exception, I should raise or sth?

Reply:
ValueError, or make a custom exception. We don't have good discipline with Exceptions right now.